rust/arrow/benches/trigonometry_kernels.rs

```diff
@@ -0,0 +1,171 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#[macro_use]
+extern crate criterion;
+use criterion::Criterion;
+
+use num::Zero;
+use rand::Rng;
+use std::{ops::Div, sync::Arc};
+
+extern crate arrow;
+
+use arrow::error::{ArrowError, Result};
+use arrow::util::test_util::seedable_rng;
+use arrow::{
+    array::*,
+    compute::*,
+    datatypes::{ArrowNumericType, ArrowPrimitiveType},
+};
+
+/// This computes the Haversine formula without using unary kernels
+fn haversine_no_unary<T>(
```

Comment:
I would try to get the result from the exercise, drop this code once the best option was found, and document which was was better in implementation notes on the decided approach.